cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -0,0 +1,90 @@
+use crate::flox::Flox;
+use std::path::Path;
+
+pub struct PublishProvider {
+    repo_dir: Path,
+    package_name: String,
+}
+
+impl PublishProvider {
+    pub fn new(repo_dir: Path, package: &str) -> Self {
+        Self { repo_dir: repo_dir, package_name: package.to_string() }
+    }
+
+    /// initiates the publish process
+    pub fn publish(&self) -> Result<(), String> {
+        // Run each phase, gathering info along the way until we get to the end
+        // and can send it all to the catalog
+
+        // Best pattern to use here?
+        // - Store in structs in Self along the way?
+        // - Type States (like the old publish?)
+        // - Make each phase independant, taking arguments for all that's needed
+        //    and track the ongoing state only here in publish()?
+        // - Combination of Type States and making each phase independant may be the 
+        //    easiest to test?
+        prepublish_check();
+
+        prepublish_build();
+
+        gather_build_info();
+
+        publish_to_catalog();
+    }
+
+    /// Ensures current commit is in remote
+    /// Gathers of the info needed for publish
+    /// ... locked_url, rev, rev_count, rev_date of the build repo from git
+    /// ... locked_url, rev, rev_count, rev_date of the base catalog from the
+    ///         lockfile of the build repo
+    /// ... version, description, name, attr_path from manifest
+    /// ... catalog from ???? (cli argument? manifest?)
+    pub fn prepublish_check(repo_dir: &Path, package_name: &str)
```

Comment:
I would do your catalog access check here.

Reply:
This method will use some kind of network client, so it will do some async stuff, but I think we can simply block on it for now and turn it into a synchronous operation. Technically you could do all of these things in parallel with async tasks, but they all seem relatively quick so it's probably not worth it given that the build in the next step is probably the bottleneck.

Reply:
Understood thanks.  And added notes to confirm access to remote resources needed.